### ISSUE TITLE
Question About PWM DIm (ESP8266)

### ISSUE THREAD

**Issue by ryancasler** [DanielOgorchock/ST_Anything#226]
Had a question for you about the EX_PWM_Dim function.  I see that it uses analog write but there is nowhere to set the frequency within the constructor.  If I were to manually add the following in the setup method after the setup of the dimmer device, would that work or would it conflict with something?

analogWriteFreq(25000);

The default frequency is only 1kHz which is far too slow for the application I'm looking at using it for.  Thanks!

**Comment by ogiewon** [comment]
Good question...  I based the PWM functionality on the official Arduino documentation for their boards.  I don’t see an analogWriteFreq() call for genuine Arduinos.  They all seem to have a constant frequency output, with variable pulse width. 

https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/

What board are you using?

**Comment by ryancasler** [comment]
A Wemos D1 mini.  The analogWriteFreq appears in the Arduino library for the ESP8266.  I confirmed to make sure I wasn't losing it.  

http://github.com/esp8266/Arduino/blob/a67986915512c5304bd7c161cf0d9c65f66e0892/cores/esp8266/core_esp8266_wiring_pwm.cpp

If I have to drop it in manually, that's fine.  I don't have a problem doing that.  Just wanted to check.  If you want to add it to the constructor for the device in the setup, that would be cool too.  

However, I'm not able to compile with the EX_PWM_Dim functionality included in the sketch.  I took the example at the top of "h" file:
`static st::EX_PWM_Dim executor1(F("dimmerSwitch1"), PIN_SWITCH_1, PIN_DIMMER_1, LOW, true);  `
But I get the following error:
`no matching function for call to 'st::EX_PWM_Dim::EX_PWM_Dim(const __FlashStringHelper*, const uint8_t&, const uint8_t&, int, bool)'`

Can you see anything that I've done incorrectly?  I haven't even added the analog frequency yet.

**Comment by ryancasler** [comment]
So, I was able to set the frequency correctly (at least I assume it is).  I am controlling a 4 wire PWM case fan and the control signal has  to be 21 - 28 kHz in order to work. After hooking everything up I am seeing exactly what I would expect.  100% duty cycle results in 100% fan speed (the same speed as without the PWM wire connected) and 0 results in the fan at it's minimum speed (this is a type A fan).  So, it definitely seems to work by putting the analogWriteFreq in the setup method.  Right now I have a separate switch device which I am tying together in Hubitat to control the relay for the 12v power line.  It would be great if we could get the switch pin tied into this device too.  I'll take a stab at it and if I get anywhere I'll send a PR but most of my Arduino knowledge is just hacking together what I'm able to <strike> steal </strike>  borrow from others. So, I wouldn't hold my breath.  But hey, you never know, right?  hehe

I'm also going to try (try being the operative word) to add an EX_PWM_Dim_8266 that has the frequency built into the device details as well.  But I have less faith in my abilities there.  Whatever I come up with, I'll send along.

As always, thanks Dan!

**Comment by vseven** [comment]
Just a FYI - This can probably be added as a optional variable and if it's not passed default to 1000 but otherwise use whatever is passed in.  It seems the 8266 code already has checks against bad values (lines 45 - 52 in linked library above).

The ESP32 also is changeable and defaults to 5000 but can go even higher then a 8266.  Couldn't find anything that does the check though for bad value but I'm sure it's in there.

**Comment by ryancasler** [comment]
> Just a FYI - This can probably be added as a optional variable

Yeah, that's what I'm going to try and do, add it in as another parameter to the device "constructor" in the sketch and therefor to the ST_Anything Libraries as well.  But as Dan might remember, this isn't the first time I've tried to do this.  Before someone finally implemented the EX_Servo library I took a crack at it too.  After 3 days of screwing around with it (and a lot of swearing) and making absolutely zero headway I finally gave up.  So, I'm not 100% sure I'll be able to do it but I'll give it a shot.  There are just so many cross references between the Executors, the EX_PWM_Dim and Everything files.  And of course there's a .h and a .cpp.  Sometimes I think they make it intentionally confusing to keep job security for the "how-to" industry.  ;-)

> The ESP32 also is changeable and defaults to 5000

And that's new for the ESP32.  There was no analogWrite function at all when the board was first launched and for quite a while after that as well.  You'll see references to it in the EX_PWM_Dim.h file.  But it looks like they've added a lot of that functionality in now.  As it stands today, the ESP32 doesn't have any PWM support in ST_Anything/Hubduino.  For the EX_RGB_Dim Dan used the ledcWrite function that was there at launch.  That could probably be updated to use analogWrite now too.  That seems to a be a pretty straightforward change so I'll try and put a PR through for that sometime today.

**Comment by vseven** [comment]
@ryancasler - I wrote the RGB library based on the PWM_Dim one if you look at the comments so I'm a little familiar.  :)    The same code using the PWM channel could be used on the EX_PWM_Dim if someone wanted to make the updates but I've only used ESP32's so I never bothered and there doesn't seem to be a demand for it.  As for your frequency issue I think you are over complicating it.  In the EX_PWM_Dim.h you'll need to add in the optional variable on line 55:

`EX_PWM_Dim(const __FlashStringHelper *name, byte pinPWM, bool startingState = LOW, bool invertLogic = false, uint16_t pwmFreq = 1000 );`

Then if you want to do it right you would add in a holder variable up around line 45:

`uint16_t m_nFreq;`

Then in the EX_PWM_Dim.CCP do the same for the header on line 127:

`EX_PWM_Dim::EX_PWM_Dim(const __FlashStringHelper *name, byte pinPWM, bool startingState, bool invertLogic, uint16_t pwmFreq) :`

Then add in the assignment to the holder variable right below it:

`m_nFreq(pwmFreq),`

And finally you'd assign it somewhere.  Maybe in the setPWMPin function around like 195?:

`analogWriteFreq(m_nFreq);`
 

This is completely just guessing....
